File: sample/client/proto/PostData.cs
//proto/Forum.proto
using ILib.ProtoPack;
using IWriter = ILib.ProtoPack.IWriter;
using IReader = ILib.ProtoPack.IReader;
using Provider = ILib.ProtoPack.InstanceProvider;

namespace Forum
{
	public partial class PostData : IMessage 
	{
		public int Id { get; set; }

		public string Message { get; set; }

		public Forum.User User { get; set; }

		/// <summary>
		/// Serialize Message
		/// </summary>
		public void Write(IWriter w, bool skipable = true)
		{
			// Write Map Length
			if (!skipable) {
				w.WriteMapHeader(3);
			} else {
				int mapLen = 0;
				if(this.Id != default(int)) mapLen++;
				if(this.Message != default(string)) mapLen++;
				if(this.User != default(Forum.User)) mapLen++;
				w.WriteMapHeader(mapLen);
			}
			
			// Write Id
			if(!skipable || this.Id != default(int)) {
				w.WriteTag(1);
				w.Write(Id);
			}
			
			// Write Message
			if(!skipable || this.Message != default(string)) {
				w.WriteTag(2);
				w.Write(Message);
			}
			
			// Write User
			if(!skipable || this.User != default(Forum.User)) {
				var User = this.User;
				w.WriteTag(3);
				if (User == default(Forum.User))
				{
					w.WriteNil();
				}
				else
				{
					User.Write(w, skipable);
				}
			}
		}

		/// <summary>
		/// Deserialize Message
		/// </summary>
		public void Read(IReader r, bool overridable = false)
		{
			// Read Map Length
			var mapLen = r.ReadMapHeader();
			uint tag = 0;
			int index = 0;

			while ((tag = r.ReadTag(index++, mapLen)) != 0)
			{
				switch(tag) {
				case 1:
					Id = r.ReadInt();
					break;
				case 2:
					Message = r.ReadString();
					break;
				case 3:
					if(r.IsNull())
					{
						User = r.ReadNil<Forum.User>();
						continue;
					}
					if(!overridable || User == default(Forum.User))
					{
						User = Provider.New<Forum.User>();
					}
					User.Read(r, overridable);
					break;
				default:
					r.Skip();
					break;
				}
			}
		}
	}
}

File: sample/server/proto/forum/postdata.js
// sample/proto/Forum.proto
"use strict";
const _packer = require('proto-msgpack');
const _proto = require('./../index.js');
class PostData {
	constructor(init, pos) {
		this.id = 0;
		this.message = null;
		if(init == null || init == true){
			this.user = new _proto.Forum.User();
		} else {
			this.user = null;
		}
		if(Buffer.isBuffer(init)){
			this.read(new _packer.ProtoReader(init, pos));
		}
	}
	pack() {
		const w = _packer.defaultWriter;
		w.clear();
		this.write(w);
		return w.toBuffer();
	}
	unpack(buf, pos) {
		if(!Buffer.isBuffer(buf)){
			this.read(buf);
		} else {
			this.read(new _packer.ProtoReader(buf, pos));
		}
	}
	write(w) {
		// Write Map Length
		w.writeMapHeader(3);
		
		// Write Id
		w.writeTag(1);
		w.writeNumber(this.id);
		
		// Write Message
		w.writeTag(2);
		w.writeString(this.message);
		
		// Write User
		w.writeTag(3);
		if (!this.user) {
			w.writeNil();
		} else {
			this.user.write(w);
		}
	}
	read(r) {
		// Read Map Length
		const mapLen = r.readMapHeader();

		for(let i = 0; i < mapLen; i++) {
			const tag = r.readTag();
			switch(tag) {
			case 1:
				this.id = r.readNumber();
				break;
			case 2:
				this.message = r.readString();
				break;
			case 3:
				if(r.isNull()) {
					r.readNil();
					this.user = null;
				} else {
					this.user = new _proto.Forum.User(false);
					this.user.read(r);
				}
				break;
			default:
				r.skip();
				break;
			}
		}
	}
}
module.exports = PostData
